WikiTreeDynamicTree.js

```diff
@@ -7,19 +7,191 @@
 		originOffsetY = 300,
 		boxWidth = 200,
 		boxHeight = 50,
+		halfBoxWidth = boxWidth / 2,
+		halfBoxHeight = boxHeight / 2,
 		nodeWidth = boxWidth * 1.5,
-		nodeHeight = boxHeight * 2;
+		nodeHeight = boxHeight * 3;
+
+	const L = -1, R = 1;
+	const ANCESTORS = 1;
+	const DESCENDANTS = -1;
+
+
+	class Couple {
+		constructor(id, a, b, focus) {
+			this.id = id;
+			if (b === undefined) {
+				if (a === undefined) {
+					throw new Error('Attempting to create an empty couple');
+				}
+				b = a.getSpouse();
+			}
+			if (focus == L || focus == R) {
+				this.focus = focus;
+			} else {
+				if (a) {
+					this.focus = L;
+				} else {
+					this.focus = R;
+				}
+			}
+			if ((focus == L && !a) || (focus == R && !b)) {
+				console.error(`Internal ERROR: The focus of a couple cannot be undefined: a=${a}, b=${b}, focus=${focus}`);
+			}
+
+			if ( (a && a.isFemale() && (!b || !b.isFemale()))
+			  || (b && b.isMale() && (!a || !a.isMale())) ) {
+				// Swap a and b
+				this.a = b;
+				this.b = a;
+				this.focus = -this.focus;
+			} else {
+				this.a = a;
+				this.b = b;
+				this.focus = this.focus;
+			}
+			condLog(`new Couple: id=${this.getId()}, focus=${this.focus}, ${this.toString()}`, this.a, this.b)
+		}
+
+		getId() {
+			return this.id;
+		}
+
+		getInFocus() {
+			if (this.focus == R) {
+				return this.b;
+			} else {
+				return this.a;
+			}
+		}
+
+		getNotInFocus() {
+			if (this.focus == R) {
+				return this.a;
+			} else {
+				return this.b;
+			}
+		}
+
+		getJointChildren() {
+			let fatherId = this.a ? this.a.getId() : undefined;
+			let motherId = this.b ? this.b.getId() : undefined;
+			let children = this.getInFocus().getChildren();
+			let list = [];
+
+			for (var i in children) {
+				var child = children[i];
+				if (child.getFatherId() == fatherId && child.getMotherId() == motherId) {
+					list.push(child);
+				}
+			}
+			return sortByBirthDate(list);
+		}
+
+		isComplete() {
+			return this.a && this.b;
+		}
+
+		hasAParent() {
+			return (this.a && this.a.hasAParent()) || (this.b && this.b.hasAParent());
+		}
+
+		isAncestorExpandable() {
+			return !this.children && this.hasAParent() &&
+				((this.a && !this.a._data.Parents) || (this.b && !this.b._data.Parents));
+		}
+
+		isDescendantExpandable() {
+			return !this.children && ((this.a && !this.a.getChildren()) || (this.b && !this.b.getChildren()));
+		}
+
+		setA(person) {
+			this.a = person;
+			return this;
+		}
+
+		setB(person) {
+			this.b = person;
+			return this;
+		}
+
+		setSpouse(person) {
+			if (person.isMale()) {

```

Comment:
This male & female coupling will imho start to be problematic. Not with historical records, but with those of the present (gays, lesbians,....), but better solve the problem when it comes...

Reply:
I don't think there is a male/female coupling at the moment (unless I missed something)?  I.e. a profile does not have to be either male or female - the code handles gender in effect as male/female/other and a couple can be any combination. It just tries to ensure consistency in  where in a couple object the male and female profiles are "placed"

Reply:
Ouch my bad...